README.md
# RF Light House (rflh)

A python script to use a rotor and a SDR device (RTL-SDR or HackRF One) to measure the RF level around and get a data set and beautiful interactive graphics.

![background noise measurement 145 Mhz](/imgs/145.png)

WARNING: This repository is new and under construction, you will see some [TODO] & "(work in progress...)" sections/docs yet.

## Motivation

This project born from a friend's challenge to measure the background noise impact on my 70cm satellite band noise floor, from a new 2/3/4G cellular tower that my ISP is setting up 50m away from my antennas and with direct sight.

Soon I realized the true potential of it and it get bigger and feature rich quickly.

## Features

At the end of the execution you get:

- A cvs file in the data folder with the resulting data
- A png image in the data folder with the rose plot

Both files are named as follows: `YYYMMDD_HHMM_device_freqMHz_BWkHz_stepo` with the matching .csv and .png extensions. The runtime text on the console name the files created for easy parsing (unless you select the 'quiet' option)

For example a real fast scan showing the references to the img & cvs file for parsing:

```
310(307.5);-102,16941998017171
320(317.5);-101,77656601734243
330(327.5);-101,55990296468812
340(337.5);-101,583492037594
350(347.5);-102,01302571365707
Scan took 0:53
CSVFile: data/20220108_1357_rtl_145.17MHz_300kHz_10o.csv
Parking the rotor in the background
Reattached kernel driver
Dynamc range: 2.9333941135273136 dB, 10%: 0.2933394113527314
Min: -104.78663648956817, Max -101.26656355333539
ImgFile: data/20220108_1357_rtl_145.17MHz_300kHz_10o.png
```

You can stop the generation of the cvs and the image files if not needed, take a peek on the options.

Also if you are on a GUI enviroment you can issue the '-i' or '--interactive' switch and at the end of the sweep a interactive matplotlib graph will popup.

For a more detailed technical stuff on the features see [OPTIONS_EXPLAINED.md](OPTIONS_EXPLAINED.md) (work in progress...)

## Installation

As any script in python you will need some dependencies, default dev env is Ubuntu Linux 20.04 LTS. I'm working/testing a single portable file for linux/windows/mac but it's not ready yet (pyinstall stuff)

The installation of the utilities & python modules are covered in the [Install](INSTALL.md) document.

At the end of the we have some examples / use cases at the end of the [OPTIONS_EXPLAINED.md](OPTIONS_EXPLAINED.md) document.

## Author, contributions, code & donations

The author is Pavel Milanes Costa (CO7WT), you can join the team contributing with code fix, improvements, bug reports, ideas, etc. Use te "Issues" tab for that.

This software is Free Software under GPLv3, see [LICENCE](LICENSE.GPLv3); free as in freedom.

If you find this piece of soft usefull and want to support the author with a tip, hardware donation or just a change for a coffee please contact me at pavelmc@gmail.com for instructions.

For money tips you can use my [QvaPay donation page](https://qvapay.com/payme/pavelmc), thanks in advance!


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

<!--
This is a note for developers about the recommended tags to keep track of the changes:

- Added: for new features.
- Changed: for changes in existing functionality.
- Deprecated: for soon-to-be removed features.
- Removed: for now removed features.
- Fixed: for any bug fixes.
- Security: in case of vulnerabilities.

Dates must be YEAR-MONTH-DAY
-->

## 2022-01-08

- Changed: Graph tool changed from plotly to pure mathplotlib.
- Removed: dark/light themes.
- Changed: Now the default is to save the csv & png file with the graph, if you have GUI on your system you can issue a '-i/--interactive' option to popup a matplotlib graph.
- Changed: Documentation to reflect the changes.

## 2022-01-07

- First release!
- Issues:
    - Incomplete documentation
    - Issues with the pyinstaller and plotly
    - Must clarify the use of rotctld

hrf.py
import sys
import numpy as np
from time import sleep
from scipy import signal
from libhackrf import *

class RF(object):
    def __init__(self, ppm):
        try:
            self.hrf = HackRF()
        except:
            print("Are you sure that there is a HackRF One connected?\nI can't find it!")
            sys.exit()
        
        # HackRF setup
        self.hrf.lna_gain = 32
        self.hrf.vga_gain = 30
        self.hrf.sample_rate = 8e6
        #self.hrf.freq_correction = ppm #TODO
        self.bw = 50
        self.f = 101500000
        self.freq = self.f - 2e6
        self.nperseg = 256
        self.bbw = 0
        self.samples = 1.024e6

        # enable/disable the built-in amplifier:
        #self.hrf.enable_amp()
        self.hrf.disable_amp()

    def set_freq(self, freq):
        '''
        freq in hz
        We put the freq in the middle of the lower half
        0----*----F----3----4
        '''

        self.freq = freq
        self.f = freq + (self.hrf.sample_rate / 4)
        self.hrf.center_freq = self.f

    def set_bw(self, bw):
        '''
        bw in khz (between 1khz and 3 Mhz)
        '''

        # failsafe
        bw = int(bw)

        # calculate the neede nperseg to get at least N samples in the bw
        if (bw < 1):
            raise ValueError("bw must be greather than 1khz")
        if (bw > 3e3):
            raise ValueError("bw must be lower than 3 Mhz")

        # N is the desired samples in the bw selected
        N = 5
        self.bw = bw * 1000
        self.nperseg = int((self.hrf.sample_rate / self.bw) * (N + 1))
        self.bbw = self.hrf.sample_rate / self.nperseg

        print("Bin bw is {} khz, with {} segments, {} samples in the BW".format(
            int(self.bbw/1e3), self.nperseg, N + 1))

    def set_gain_lna(self, gain):
        self.hrf.lna_gain = gain

    def set_gain_vga(self, gain):
        self.hrf.vga_gain = gain

    def amp_on(self):
        self.hrf.enable_amp()

    def amp_off(self):
        self.hrf.disable_amp()

    def get_average(self):

        # get the samples
        samples = self.hrf.read_samples(self.samples)
        freqs, Pxx = signal.welch(
            samples, fs=self.hrf.sample_rate, nperseg=self.nperseg, return_onesided=False)

        # # use matplotlib to estimate and plot the PSD
        # psd(samples, NFFT=8192, Fs=self.hrf.sample_rate /
        #     1e6, Fc=self.hrf.center_freq/1e6)
        # xlabel('Frequency (MHz)')
        # ylabel('Relative power (dB)')
        # show()

        # Shift frequencies by the center frequency during sample stage
        freqs += self.f

        # Use the 'power' formula for dB (10*log10(X))
        # The 'np.abs(Pxx)' is there because 'Pxx' is complex-valued
        adB = 10 * np.log10(np.abs(Pxx))

        # Parse the arrays to get just the needed ones, amplitude and samples
        ampsum = 0
        samples = 0
        index = 0

        start = self.freq - (self.bw/2)
        stop = self.freq + (self.bw/2)
        for fs in freqs:
            if (fs >= start and fs <= stop):
                ampsum += adB[index]
                samples += 1
            index += 1

        # calc mean
        average_db = ampsum / samples
        return average_db

    def fast(self):
        # do a fast scan, just 512k samples
        self.bucket(512e3)

    def bucket(self, s):
        # set the sampling bucket size
        self.samples = s

    def close(self):
        # release the hackrf
        self.hrf.close()

if __name__ == "__main__":
    hrf = RF(0)
    hrf.set_freq(446000000)
    hrf.set_bw(50)
    man_level = hrf.get_average()
    # debug
    print("Frequency: 440. MHz, 200 Khz bandwidth")
    print("Mean level: {}".format(man_level))



mpl.py
import numpy as np

import matplotlib.pyplot as plt
from matplotlib.patches import Circle, RegularPolygon
from matplotlib.path import Path
from matplotlib.projections.polar import PolarAxes
from matplotlib.projections import register_projection
from matplotlib.spines import Spine
from matplotlib.transforms import Affine2D


def radar_factory(num_vars, frame='circle'):
    """
    Create a radar chart with `num_vars` axes.

    This function creates a RadarAxes projection and registers it.

    Parameters
    ----------
    num_vars : int
        Number of variables for radar chart.
    frame : {'circle', 'polygon'}
        Shape of frame surrounding axes.

    """
    # calculate evenly-spaced axis angles
    theta = np.linspace(0, 2*np.pi, num_vars, endpoint=False)

    class RadarAxes(PolarAxes):

        name = 'radar'
        # use 1 line segment to connect specified points
        RESOLUTION = 1

        def __init__(self, *args, **kwargs):
            super().__init__(*args, **kwargs)
            # rotate plot such that the first axis is at the top
            self.set_theta_zero_location('N')

        def fill(self, *args, closed=True, **kwargs):
            """Override fill so that line is closed by default"""
            return super().fill(closed=closed, *args, **kwargs)

        def plot(self, *args, **kwargs):
            """Override plot so that line is closed by default"""
            lines = super().plot(*args, **kwargs)
            for line in lines:
                self._close_line(line)

        def _close_line(self, line):
            x, y = line.get_data()
            # FIXME: markers at x[0], y[0] get doubled-up
            if x[0] != x[-1]:
                x = np.append(x, x[0])
                y = np.append(y, y[0])
                line.set_data(x, y)

        def set_varlabels(self, labels):
            self.set_thetagrids(np.degrees(theta), labels)

        def _gen_axes_patch(self):
            # The Axes patch must be centered at (0.5, 0.5) and of radius 0.5
            # in axes coordinates.
            if frame == 'circle':
                return Circle((0.5, 0.5), 0.5)
            elif frame == 'polygon':
                return RegularPolygon((0.5, 0.5), num_vars,
                                      radius=.5, edgecolor="k")
            else:
                raise ValueError("Unknown value for 'frame': %s" % frame)

        def _gen_axes_spines(self):
            if frame == 'circle':
                return super()._gen_axes_spines()
            elif frame == 'polygon':
                # spine_type must be 'left'/'right'/'top'/'bottom'/'circle'.
                spine = Spine(axes=self,
                              spine_type='circle',
                              path=Path.unit_regular_polygon(num_vars))
                # unit_regular_polygon gives a polygon of radius 1 centered at
                # (0, 0) but we want a polygon of radius 0.5 centered at (0.5,
                # 0.5) in axes coordinates.
                spine.set_transform(Affine2D().scale(.5).translate(.5, .5)
                                    + self.transAxes)
                return {'polar': spine}
            else:
                raise ValueError("Unknown value for 'frame': %s" % frame)

    register_projection(RadarAxes)
    return theta


def example_data():
    # The following data is from the Denver Aerosol Sources and Health study.
    # See doi:10.1016/j.atmosenv.2008.12.017
    #
    # The data are pollution source profile estimates for five modeled
    # pollution sources (e.g., cars, wood-burning, etc) that emit 7-9 chemical
    # species. The radar charts are experimented with here to see if we can
    # nicely visualize how the modeled source profiles change across four
    # scenarios:
    #  1) No gas-phase species present, just seven particulate counts on
    #     Sulfate
    #     Nitrate
    #     Elemental Carbon (EC)
    #     Organic Carbon fraction 1 (OC)
    #     Organic Carbon fraction 2 (OC2)
    #     Organic Carbon fraction 3 (OC3)
    #     Pyrolized Organic Carbon (OP)
    #  2)Inclusion of gas-phase specie carbon monoxide (CO)
    #  3)Inclusion of gas-phase specie ozone (O3).
    #  4)Inclusion of both gas-phase species is present...
    data = [
        ['Sulfate', 'Nitrate', 'EC', 'OC1', 'OC2', 'OC3', 'OP', 'CO', 'O3'],
        ('Basecase', [
            [0.88, 0.01, 0.03, 0.03, 0.00, 0.06, 0.01, 0.00, 0.00],
            [0.07, 0.95, 0.04, 0.05, 0.00, 0.02, 0.01, 0.00, 0.00],
            [0.01, 0.02, 0.85, 0.19, 0.05, 0.10, 0.00, 0.00, 0.00],
            [0.02, 0.01, 0.07, 0.01, 0.21, 0.12, 0.98, 0.00, 0.00],
            [0.01, 0.01, 0.02, 0.71, 0.74, 0.70, 0.00, 0.00, 0.00]])
    ]
    return data


if __name__ == '__main__':
    N = 9
    theta = radar_factory(N)

    data = example_data()
    spoke_labels = data.pop(0)

    fig, axs = plt.subplots(figsize=(9, 9), nrows=2, ncols=2,
                            subplot_kw=dict(projection='radar'))
    fig.subplots_adjust(wspace=0.25, hspace=0.20, top=0.85, bottom=0.05)

    colors = ['b', 'r', 'g', 'm', 'y']
    # Plot the four cases from the example data on separate axes
    for ax, (title, case_data) in zip(axs.flat, data):
        ax.set_rgrids([0.2, 0.4, 0.6, 0.8])
        ax.set_title(title, weight='bold', size='medium', position=(0.5, 1.1),
                     horizontalalignment='center', verticalalignment='center')
        for d, color in zip(case_data, colors):
            ax.plot(theta, d, color=color)
            ax.fill(theta, d, facecolor=color, alpha=0.25)
        ax.set_varlabels(spoke_labels)

    # add legend relative to top-left plot
    labels = ('Factor 1')
    legend = axs[0, 0].legend(labels, loc=(0.9, .95),
                              labelspacing=0.1, fontsize='small')

    fig.text(0.5, 0.965, '5-Factor Solution Profiles Across Four Scenarios',
             horizontalalignment='center', color='black', weight='bold',
             size='large')

    plt.show()


rtl.py
import sys
import numpy as np
from time import sleep
from scipy import signal
from rtlsdr import RtlSdr

class RF(object):
    def __init__(self, ppm):
        # rtl-sdr
        try:
            self.rtl = RtlSdr()
        except:
            print("Are you sure that there is a HackRF One connected?\nI can't find it!")
            sys.exit()

        # setup
        self.rtl.gain = 28.0
        # Warn! samples must be 2^n number, aka 2048 instead of 2k, 512 instead of 500
        self.rtl.sample_rate = 2.048e6
        if ppm != 0:
            self.rtl.freq_correction = ppm
        self.bw = 50
        self.f = 101500000
        self.freq = self.f - 0.512e6
        self.nperseg = 256
        self.bbw = 0
        self.samples = 1.024e6

    def set_freq(self, freq):
        '''
        freq in hz
        We put the freq in the middle of the lower half
        0----*----F----3----4
        '''

        self.freq = freq
        self.f = freq + (self.rtl.sample_rate / 4)
        self.rtl.center_freq = self.f

    def set_bw(self, bw):
        '''
        bw in khz (between 1khz and 3 Mhz)
        '''

        # failsafe
        bw = int(bw)

        # calculate the neede nperseg to get at least N samples in the bw
        if (bw < 1):
            raise ValueError("bw must be greather than 1kHz")
        if (bw > 800e3):
            raise ValueError("bw must be lower than 800 kHz")

        # N is the desired samples in the bw selected
        N = 5
        self.bw = bw * 1000
        self.nperseg = int((self.rtl.sample_rate / self.bw) * (N + 1))
        self.bbw = self.rtl.sample_rate / self.nperseg

        print("Bin bw is {} khz, with {} segments, {} samples in the BW".format(
            int(self.bbw/1e3), self.nperseg, N + 1))

    def set_gain_lna(self, gain):
        self.rtl.gain = gain

    # compat option against hackrf 
    def set_gain_vga(self, gain):
        return True

    # compat option against hackrf
    def amp_on(self):
        return True

    # compat option against hackrf
    def amp_off(self):
        return True

    def get_average(self):

        # get the samples
        samples = self.rtl.read_samples(self.samples)
        freqs, Pxx = signal.welch(
            samples, fs=self.rtl.sample_rate, nperseg=self.nperseg, return_onesided=False)

        # # use matplotlib to estimate and plot the PSD
        # psd(samples, NFFT=8192, Fs=self.rtl.sample_rate /
        #     1e6, Fc=self.rtl.center_freq/1e6)
        # xlabel('Frequency (MHz)')
        # ylabel('Relative power (dB)')
        # show()

        # Shift frequencies by the center frequency during sample stage
        freqs += self.f

        # Use the 'power' formula for dB (10*log10(X))
        # The 'np.abs(Pxx)' is there because 'Pxx' is complex-valued
        adB = 10 * np.log10(np.abs(Pxx))

        # Parse the arrays to get just the needed ones, amplitude and samples
        ampsum = 0
        samples = 0
        index = 0

        start = self.freq - (self.bw/2)
        stop = self.freq + (self.bw/2)
        for fs in freqs:
            if (fs >= start and fs <= stop):
                ampsum += adB[index]
                samples += 1
            index += 1

        # calc mean
        average_db = ampsum / samples
        return average_db

    def bucket(self, s):
        # set the sampling bucket size
        self.samples = s

    def fast(self):
        # set the sampling bucket size
        self.bucket(512e3)

    def close(self):
        # release the hackrf
        self.rtl.close()

if __name__ == "__main__":
    rtl = RF(68)
    rtl.set_freq(145000000)
    rtl.set_bw(50)
    rtl.fast()
    man_level = rtl.get_average()
    # debug
    print("Frequency: 440. MHz, 200 Khz bandwidth")
    print("Mean level: {}".format(man_level))



rotor.py
import sys
import os
import Hamlib
import time
import subprocess
from pathlib import Path
from configparser import ConfigParser

# default vars
MINSTEP = 5.0
configfile = 'rotor.conf'

class Rotor():
    def __init__(self):
        # load config and start the sub-process control if needed
        self.loadconfig()
        self.start_rotor()

        # limits
        self.MIN_AZ = 0
        self.MIN_EL = 0
        self.MAX_AZ = 360
        self.MAX_EL = 90

    def loadconfig(self):
        # find the correct place for the config file
        local = True
        cwd = os.getcwd()
        defcf = os.path.join(cwd, configfile)

        # load config from the rotor.conf file
        config = ConfigParser()
        if os.path.exists(defcf) and os.path.isfile(defcf):
            print("rotor.conf at: {}".format(defcf))
            config.read(defcf)
        else:
            if local:
                print("Can't find the '{}' in the local directory".format(configfile))
                print("Please copy & edit the rotor.conf file from the project page to this directory.")
                sys.exit()
            else:
                import shutils
                shutil.copyfile(os.path.join(sys._MEIPASS, configfile), defcf)
                print("No rotor config file, a new one has been created, please edit it with your particular rotor configs and try it again.")
                print(defcf)
                sys.exit()

        self.default = config.get('DEFAULT', 'rotor')
        self.model = config.get(self.default, 'model')
        self.device = config.get(self.default, 'device')
        self.options = config.get(self.default, 'options')

    def start_rotor(self):
        # Disable all debug output from Hamlib
        Hamlib.rig_set_debug(Hamlib.RIG_DEBUG_NONE)

        # Create rotor object of type net
        # self.rot = Hamlib.Rot(Hamlib.ROT_MODEL_NETROTCTL)
        self.rot = Hamlib.Rot(int(self.model))

        # Setup device
        self.rot.set_conf("rot_pathname", self.device)

        # setup extra options 
        opts = self.options.strip()
        if len(opts) > 0:
            if not (len(opt) % 2):
                print("Strange, options must be in pairs, trying but may fail...")

            pairs = []
            cmds = []
            s = opts.split(' ')
            for i in range(len(s)/2):
                cmd = s[i*2]
                data = s[i*2 + 1]
                if cmd == '-s':
                    cmds.append(['serial-speed', data])
                if cmd == '-C':
                    cmds.append([data.split('=')[0],
                                data.split('=')[1]])

            for (opt, value) in cmds:
                self.rot.set_conf(opt, value)

        # Open rotor
        # The Python bindings for Hamlib does not return anything
        # so we have no knowledge if this was actually successful...
        self.rot.open()

        # get the limits from the rotor
        self.MIN_AZ = float(self.rot.get_conf("min_az"))
        self.MIN_EL = float(self.rot.get_conf("min_el"))
        self.MAX_AZ = float(self.rot.get_conf("max_az"))
        self.MAX_EL = float(self.rot.get_conf("max_el"))

    def set_position(self, az, elv):
        # check limits
        az = min(self.MAX_AZ, max(self.MIN_AZ, az))
        elv = min(self.MAX_EL, max(self.MIN_EL, elv))

        # ask postion
        [aaz, ael] = self.get_position()
        diff = az - aaz
        if abs(diff) <= MINSTEP:
            self.rot.set_position(aaz + (2 * diff), ael)
            time.sleep(0.25)
        self.rot.set_position(az, elv)

        while (abs(aaz - az) > 1):
            time.sleep(1)
            [aaz, ael] = self.get_position()

    def get_position(self):
        pos = self.rot.get_position()
        return pos

    def go_to(self, az, elv=0):
        self.rot.set_position(az, elv)

    def close(self):
        # close all connections
        if  self.rot:
            self.rot.close()

if __name__ == "__main__":
    r = Rotor()
    a = 0
    for az in range(0, 90, 10):
        r.set_position(az, 0)
        (a, e) = r.get_position()
        print("Azimuth, actual: {}, set: {}".format(a, az))

    # park
    r.go_to(0, 0)
    r.close()


rflh.py
import time
import argparse
import os
import matplotlib.pyplot as plt
import random as rnd
from math import pi
from rotor import *

# description
parser = argparse.ArgumentParser(description='Sweep the 360o around recording the signals levels of a frequency at a given bandwidth; using rotctld and a RTL-SDR (default) or a HackRF One for spectrum sensing.')

# frequency
parser.add_argument(
    'frequency',
    type=float,
    help='The center frequency to listen to, in MHz, "145.170" for 145.170 MHz')

# -b / --bandwidth
parser.add_argument(
    '-b ', '--bandwidth',
    type=int,
    help='The bandwidth to sample in kHz: 1-800 kHz using the RTL-SDR or 1kHz-MHz using the HackRF, 20kHz by default')

# -s / --step
parser.add_argument(
    '-s', '--step',
    type=int,
    help='Azimuth steps in degrees, 10 degrees by default, see project README.md for details.')

# -u / --paused
parser.add_argument(
    '-u', '--paused',
    help='Sweep mode: paused mode means move to target azimuth & take a measurement the repeat it (slow). The default is to make one full turn and do measurements on the fly (faster but may fail, read the documentation)',
    action="store_true")

# -bs / --bucketsize
parser.add_argument(
    '-t', '--bucketsize',
    type=int,
    help='Bucket size: how many bytes to get for processing at each sample time, 1.024 Mbytes by default (use 2^n units or it will fail) 1.024 Mbytes = 1024000')

# -p / --ppm
parser.add_argument(
    '-p', '--ppm',
    type=int,
    help='Frequency correction for you device, by default 0.0')

# -q / --quiet
parser.add_argument(
    '-q', '--quiet',
    help='Quiet: supress output, default: verbose',
    action="store_true")

# -j / --just_data
parser.add_argument(
    '-j', '--just_data',
    help='Don\'t generate & save the graph, default: generate & save it',
    action="store_true")

# -n / --nofile
parser.add_argument(
    '-n', '--nofile',
    help='Don\'t save the csv file, default: save it',
    action="store_true")

# -i / --interactive
parser.add_argument(
    '-i', '--interactive',
    help='Pop up a Matplotlib interactive graph with the results, default is no pop up',
    action="store_true")

# -d / --dummy
parser.add_argument(
    '-d', '--dummy',
    help='Don\'t use the rotor or the RTL-SDR or HackRF, just generate a dummy dataset and plot it; it implies --nofile; just for testing purposes',
    action="store_true")

# -l / --lna
parser.add_argument(
    '-l', '--lna',
    type=float,
    help='LNA gain value: 0-49.6 dB in 0.4 dB for the RTL-SDR or 0-40 dB in 8 dB steps for HackRF. Defaults: 28.0 for the RTL-SDR & 32 for the HackRF')

# -v / --vga
parser.add_argument(
    '-v', '--vga',
    type=int,
    help='VGA gain value (Only HackRF One): 0-62 in 2 units steps, 30 by default')

# -a / --amp_on
parser.add_argument(
    '-a', '--amp_on',
    help='Amplifier on (Only HackRF One): by default it\'s disabled; WATCH OUT! some firmware revision has this option reversed (mine has it)',
    action="store_true")

# -o / --hackrfone
parser.add_argument(
    '-o', '--hackrfone',
    help='Use a HackRF One instead the default: RTL-SDR',
    action="store_true")

args = parser.parse_args()

# args setup
fq = float(args.frequency)
bw = 20
if args.bandwidth:
    bw = args.bandwidth

if args.step:
    if args.step < 6:
        print("WARNING: You selected a azimuth step lower/equal than 6 degrees: most rotors can't handle that")
    astep = args.step
else:
    astep = 10

just_data = False
if args.just_data:
    just_data = True

quiet = False
if args.quiet:
    quiet = True

nofile = False
if args.nofile:
    nofile = True

interactive = False
if args.interactive:
    interactive = True

dummy = False
if args.dummy:
    dummy = True

lna_gain = None
if args.lna:
    lna_gain = args.lna

vga_gain = None
if args.vga:
    vga_gain = args.vga

amp_on = False
if args.amp_on:
    amp_on = True

hackrfone = False
if args.hackrfone:
    hackrfone = True

ppm = 0
if args.ppm:
    ppm = args.ppm

# bucketsize
bucketsize = 1.024e6
if args.bucketsize:
    bucketsize = args.bucketsize

paused = False
if args.paused:
    paused = True
elif args.step and args.step <= 5:
    print("WARNING: You selected a step less than 6 degrees & fast scanning, most hardware can\'t handle that!")

# conditional load of the device
if hackrfone:
    from hrf import *
    if not quiet:
        print("Using HackRF One as RF device")
else:
    from rtl import *
    if not quiet:
        print("Using RTL-SDR as RF device")

# instantiating if not testing
if not dummy:
    r = Rotor()
    rf = RF(ppm)

device = 'rtl'
if hackrfone:
    device = 'hackrf'

# array that will hold the data
labels = []
levels = []
f = fq * 1000000
duration = 0

def clean_house():
    if not dummy:
        rf.close()
        r.go_to(0, 0)

# All is wraped to detect Ctrl+c
try:
    # RF setup & rotor parking
    if not dummy:
        # set HackRF
        rf.set_freq(f)
        rf.set_bw(bw)
        rf.bucket(bucketsize)
        if vga_gain != None and hackrfone:
            rf.set_gain_vga(vga_gain)
            if not quiet:
                print("VGA gain set to {}".format(vga_gain))
        if lna_gain != None:
            rf.set_gain_lna(lna_gain)
            if not quiet:
                print("LNA gain set to {}".format(lna_gain))
        if amp_on and hackrfone:
            rf.amp_on()
            if not quiet:
                print("Amp turned on!")

        # rotor parking advice
        if not quiet:
            print("Parking the rotor, please wait...")

        # rotor parking (blocking)
        r.set_position(0, 0)

    # rotor parking advice
    if not quiet:
        if not dummy:
            print("Parking done, starting the sweep")
            print("Sweep for {} Hz ({} MHz) with {} kHz of bandwidth & {} degrees of step.".format(
            f, f/1e6, bw, astep))

    # getting time for the file
    dt = time.strftime("%Y%m%d_%H%M")

    # if dummy data
    if not dummy:
        if paused:
            # slow scanning
            tstart = time.time()
            for p in range(0, 360, astep):
                labels.append(str(p))
                r.set_position(p, 0)
                l = rf.get_average()
                levels.append(l)
                if not quiet:
                    print("{};{}".format(p, str(l).replace(".", ",")))
            
            tstop = time.time()
        else:
            # fast scanning
            a = 0
            at = astep
            tstart = time.time()
            # zero
            labels.append(str(a))
            l = rf.get_average()
            levels.append(l)
            if not quiet:
                print("0(0);{}".format(str(l).replace(".", ",")))
            # start turning!
            r.go_to(360)
            while a < (360 - astep):
                (a, e) = r.get_position()
                if (at - a) < (astep / 2):
                    # data 
                    l = rf.get_average()
                    levels.append(l)
                    (an, e) = r.get_position()
                    am = (an + a) / 2
                    labels.append(str(at))
                    # debug
                    if not quiet:
                        print("{}({});{}".format(at, am, str(l).replace(".", ",")))
                        if (am - at) >= astep:
                            print("ERROR!\n\nThe selected azimuth step is to small or bucket size to big, please adjust them and try again!")
                            clean_house()
                            sys.exit()

                    # increment at
                    at += astep

            tstop = time.time()

        # time elapsed 
        duration = tstop - tstart
        if not quiet:
            print("Scan took {}:{}".format(int(duration/60), int(duration % 60)))

        # Create the name for the file if not told otherwise
        if not nofile:
            dfolder = os.path.join(os.getcwd(), 'data')
            # the data folder exists?
            if not os.path.exists(dfolder):
                os.mkdir(dfolder)
            else:
                if not os.path.isdir(dfolder):
                    os.unlink(dfolder)
                    os.path.mkdir(dfolder)

            # 20211219_2221_rtl_436.5MHz_200kHz_10o.csv
            savefile = "{}_{}_{}MHz_{}kHz_{}o".format(dt, device, fq, bw, astep)
            with open(os.path.join(dfolder, savefile + '.csv'), 'w') as f:
                # write header
                f.writelines("Degrees;dBFs\n")
                i = 0
                for v in labels:
                    f.writelines(str(v) + ";" + str(levels[i]).replace(".", ",") + "\n")
                    i = i + 1

            if not quiet:
                print("CSVFile: data/{}".format(savefile + ".csv"))
    else:
        # fake data
        rnd.seed()
        astep = 5
        units = 360/astep
        labels = [x for x in range(0, 360, astep)]
        start = rnd.randint(-1100, -600)
        stop = rnd.randint(start, 50)
        levels = []
        for l in labels:
            levels.append(rnd.randrange(start, stop)/10.0)

    if not quiet and not dummy:
        print("Parking the rotor in the background")
        clean_house()

    # statistics
    margin = 0.5
    amin = min(levels)
    amax = max(levels)
    tmargin = abs(amax - amin)
    if tmargin < 1:
        margin = 0.5
    else:
        margin = tmargin * 0.1
    lmin = amin - margin
    lmax = amax + margin

    if not quiet:
        print("Dynamc range: {} dB, 10%: {}".format(tmargin, margin))
        print("Min: {}, Max {}".format(lmin, lmax))

    if not just_data:
        speed = 'fast'
        if paused:
            speed = 'paused'

        title = u"{}: ({}: {}) {:.3f} MHz, BW: {:.1f} kHz,\n{}o steps, {:.1f} dB of DNR".format(
            device.upper(),
            speed,
            "{}:{} min".format(
                int(duration / 60),
                int(duration % 60)
            ),
            fq,
            bw,
            astep,
            tmargin
        )

        # Turn interactive plotting off
        plt.ioff()

        # create
        fig = plt.figure()
        ax = fig.add_subplot(label='title')
        fig.subplots_adjust(top=0.85)

        # Set titles for the figure and the subplot respectively
        fig.suptitle(title, fontsize=12, fontweight='normal')

        # repeat the last value to close the plot
        levels += levels[:1]

        # how many ticks/labels on the plot
        N = len(labels)

        # What will be the angle of each axis in the plot? (we divide the plot / number of variable)
        angles = [n / float(N) * 2 * pi for n in range(N)]
        # repeat the first one to close the graph
        angles += angles[:1]

        # Initialise the spider plot
        ax = plt.subplot(111, polar=True, label='graph')

        # Draw one axe per variable + add labels
        plt.xticks(angles, labels, color='grey', size=8)

        # set labels limits
        plt.ylim(lmin, lmax)

        # Plot data
        ax.plot(angles, levels, linewidth=1, linestyle='solid')

        # rotate
        ax.set_theta_zero_location('N')

        # make it clockwise
        ax.set_theta_direction(-1)

        # Fill area
        ax.fill(angles, levels, 'b', alpha=0.1)

        # save only if not dummy
        if not dummy and not just_data:
            plt.savefig(
                os.path.join(dfolder, savefile + '.png'),
                bbox_inches='tight'
            )
            if not quiet:
                print("ImgFile: data/{}".format(savefile + ".png"))

        # Show the graph if instructed to
        if interactive:
            plt.show()

except KeyboardInterrupt:
    print("\n\nCatching Ctrl+C: cleaning the house before leaving...")
    clean_house()
    sys.exit()
